File: packages/helpers/src/compliancePortalVisibility.ts
// Copyright (c) 2025-2026 Probo Inc <hello@probo.com>.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

type Translator = (s: string) => string;

export type CompliancePortalVisibility = "NONE" | "RESTRICTED" | "PUBLIC";

export const compliancePortalVisibilities = [
  "NONE",
  "RESTRICTED",
  "PUBLIC",
] as const;

export const getCompliancePortalVisibilityVariant = (visibility: CompliancePortalVisibility) => {
  switch (visibility) {
    case "NONE":
      return "danger" as const;
    case "RESTRICTED":
      return "warning" as const;
    case "PUBLIC":
      return "success" as const;
    default:
      return "neutral" as const;
  }
};

export const getCompliancePortalVisibilityLabel = (visibility: CompliancePortalVisibility) => {
  switch (visibility) {
    case "NONE":
      return "None";
    case "RESTRICTED":
      return "Restricted";
    case "PUBLIC":
      return "Public";
    default:
      return visibility;
  }
};

export type CompliancePortalLinkedVisibility = "RESTRICTED" | "PUBLIC";

export const compliancePortalLinkedVisibilities = [
  "RESTRICTED",
  "PUBLIC",
] as const;

export function getCompliancePortalLinkedVisibilityOptions(t: Translator) {
  return compliancePortalLinkedVisibilities.map((visibility) => ({
    value: visibility,
    label: t({
      "RESTRICTED": "helpers.compliancePortalVisibility.restricted",
      "PUBLIC": "helpers.compliancePortalVisibility.public",
    }[visibility]),
    variant: getCompliancePortalVisibilityVariant(visibility),
  }));
}

export function getCompliancePortalVisibilityOptions(t: Translator) {
  return compliancePortalVisibilities.map((visibility) => ({
    value: visibility,
    label: t({
      "NONE": "helpers.compliancePortalVisibility.none",
      "RESTRICTED": "helpers.compliancePortalVisibility.restricted",
      "PUBLIC": "helpers.compliancePortalVisibility.public",
    }[visibility]),
    variant: getCompliancePortalVisibilityVariant(visibility),
  }));
}

File: pkg/coredata/compliance_portal_visibility.go
// Copyright (c) 2025-2026 Probo Inc <hello@probo.com>.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package coredata

import (
	"encoding"
	"fmt"
)

type CompliancePortalVisibility string

const (
	CompliancePortalVisibilityNone       CompliancePortalVisibility = "NONE"
	CompliancePortalVisibilityRestricted CompliancePortalVisibility = "RESTRICTED"
	CompliancePortalVisibilityPublic     CompliancePortalVisibility = "PUBLIC"
)

var (
	_ fmt.Stringer             = CompliancePortalVisibility("")
	_ encoding.TextMarshaler   = CompliancePortalVisibility("")
	_ encoding.TextUnmarshaler = (*CompliancePortalVisibility)(nil)
)

func CompliancePortalVisibilities() []CompliancePortalVisibility {
	return []CompliancePortalVisibility{
		CompliancePortalVisibilityNone,
		CompliancePortalVisibilityRestricted,
		CompliancePortalVisibilityPublic,
	}
}

// compliancePortalVisibilityStrings renders visibilities for a
// trust_center_visibility[] bind. It returns nil for an empty slice so queries
// can treat NULL as "no visibility restriction".
func compliancePortalVisibilityStrings(visibilities []CompliancePortalVisibility) []string {
	if len(visibilities) == 0 {
		return nil
	}

	out := make([]string, len(visibilities))
	for i, v := range visibilities {
		out[i] = v.String()
	}

	return out
}

func (v CompliancePortalVisibility) IsValid() bool {
	switch v {
	case
		CompliancePortalVisibilityNone,
		CompliancePortalVisibilityRestricted,
		CompliancePortalVisibilityPublic:
		return true
	}

	return false
}

func (v CompliancePortalVisibility) String() string {
	return string(v)
}

func (v CompliancePortalVisibility) MarshalText() ([]byte, error) {
	return []byte(v.String()), nil
}

func (v *CompliancePortalVisibility) UnmarshalText(text []byte) error {
	val := CompliancePortalVisibility(text)
	if !val.IsValid() {
		return fmt.Errorf("invalid CompliancePortalVisibility value: %q", string(text))
	}

	*v = val

	return nil
}
